experimental/parser/parse_expr.go

```diff
@@ -0,0 +1,211 @@
+					//
+					// In other words, consider the following, inside of an expression
+					// context:
+					//
+					// foo: bar { ... }
+					//
+					// We want to diagnose the { as unexpected here, and it is better
+					// for that to be done by whatever is calling parseExpr since it
+					// will have more context.
+					return p.NewExprKV(ast.ExprFieldArgs{
+						Key:   lhs,
+						Value: parseExprInfix(p, c, where, ast.ExprAny{}, prec+1),
+					}).AsAny()
+				}
+			}
+		}
+
+		return parseExprInfix(p, c, where, lhs, prec+1)
+
+	case 1:
+		//nolint:gocritic // This is a switch for consistency with the rest of the file.
+		switch next.Text() {
+		case "to":
+			return p.NewExprRange(ast.ExprRangeArgs{
+				Start: lhs,
+				To:    c.Pop(),
+				End:   parseExprInfix(p, c, taxa.KeywordTo.After(), ast.ExprAny{}, prec),
+			}).AsAny()
+		}
+
+		return parseExprInfix(p, c, where, lhs, prec+1)
+
+	default:
+		return lhs
+	}
+}
+
+// parseExprInfix parses a prefix expression.
+//
+// This is separate from "solo" expressions because if we every gain suffix-type
+// expressions, such as f(), we need to parse -f() as -(f()), not (-f)().
+func parseExprPrefix(p *parser, c *token.Cursor, where taxa.Place) ast.ExprAny {
+	next := peekTokenExpr(p, c)
+	switch {
+	case next.Nil():
+		return ast.ExprAny{}
+
+	case next.Text() == "-":
+		c.Pop()
+		inner := parseExprPrefix(p, c, taxa.Minus.After())
+		return p.NewExprPrefixed(ast.ExprPrefixedArgs{
+			Prefix: next,
+			Expr:   inner,
+		}).AsAny()
+
+	default:
+		return parseExprSolo(p, c, where)
+	}
+}
+
+// ParseExpr attempts to parse an "solo" expression, which is an expression that
+// does not contain any infix operators.
+//
+// May return nil if parsing completely fails.
+func parseExprSolo(p *parser, c *token.Cursor, where taxa.Place) ast.ExprAny {
+	next := peekTokenExpr(p, c)
+	switch {
+	case next.Nil():
+		return ast.ExprAny{}
+
+	case next.Kind() == token.String, next.Kind() == token.Number:
+		return ast.ExprLiteral{Token: c.Pop()}.AsAny()
+
+	case canStartPath(next):
+		return ast.ExprPath{Path: parsePath(p, c)}.AsAny()
+
+	case (next.Text() == "{" || next.Text() == "[") && !next.IsLeaf():
+		body := c.Pop()
+		in := taxa.Dict
+		if next.Text() == "[" {
+			in = taxa.Array
+		}
+
+		elems := commas(p, body.Children(), true, func(c *token.Cursor) (ast.ExprAny, bool) {
+			expr := parseExpr(p, c, in.In())
+			return expr, !expr.Nil()
+		})
+
+		if next.Text() == "[" {
+			array := p.NewExprArray(body)
+			elems(func(expr ast.ExprAny, comma token.Token) bool {
+				array.AppendComma(expr, comma)
+				return true
+			})
+			return array.AsAny()
+		}
+
+		dict := p.NewExprDict(body)
+		elems(func(expr ast.ExprAny, comma token.Token) bool {
+			field := expr.AsField()
+			if field.Nil() {
+				p.Error(errUnexpected{
+					what:  expr,
+					where: in.In(),
+					want:  taxa.DictField.AsSet(),
+				})
+
+				field = p.NewExprKV(ast.ExprFieldArgs{Value: expr})
```

Comment:
I thought we decided to name this method `NewExprField`.

Reply:
I'm pretty sure we did. I think it got lost in a diff somewhere.